lib/smack/Regions.cpp

```diff
@@ -18,111 +13,27 @@ namespace smack {
-    return A->getParent();
-  else if (const BasicBlock *BB = dyn_cast<BasicBlock>(V))
-    return BB->getParent();
-
-  // XXX I know this looks bad, but it works for now
-  for (auto U : V->users())
-    return getFunction(U);
-
-  llvm_unreachable("Unexpected value.");
-}
-
-bool isFieldDisjoint(DSAWrapper *DSA, const Value *V, unsigned offset) {
-  if (const GlobalValue *G = dyn_cast<GlobalValue>(V))
-    return DSA->isFieldDisjoint(G, offset);
-  else
-    return DSA->isFieldDisjoint(V, getFunction(V));
-}
-} // namespace
 
 void Region::init(Module &M, Pass &P) {
   DL = &M.getDataLayout();
   DSA = &P.getAnalysis<DSAWrapper>();
 }
 
-namespace {
-unsigned numGlobals(const DSNode *N) {
-  unsigned count = 0;
-
-  // shamelessly ripped from getCaption(..) in lib/DSA/Printer.cpp
-  EquivalenceClasses<const GlobalValue *> *GlobalECs = 0;
-  const DSGraph *G = N->getParentGraph();
-  if (G)
-    GlobalECs = &G->getGlobalECs();
-
-  for (auto i = N->globals_begin(), e = N->globals_end(); i != e; ++i) {
-    count += 1;
-
-    if (GlobalECs) {
-      // Figure out how many globals are equivalent to this one.
-      auto I = GlobalECs->findValue(*i);
-      if (I != GlobalECs->end()) {
-        count +=
-            std::distance(GlobalECs->member_begin(I), GlobalECs->member_end()) -
-            1;
-      }
-    }
-  }
+bool Region::isSingleton(const Value *v, unsigned length) {
+  // TODO can we do something for non-global nodes?
+  auto node = DSA->getNode(v);
 
-  return count;
-}
-} // namespace
-
-bool Region::isSingleton(const DSNode *N, unsigned offset, unsigned length) {
-  if (N->isGlobalNode() && numGlobals(N) == 1 && !N->isArrayNode() &&
-      !N->isAllocaNode() && !N->isHeapNode() && !N->isExternalNode() &&
-      !N->isUnknownNode()) {
-
-    // TODO can we do something for non-global nodes?
-
-    // TODO don’t need to know if there are other members of this class, right?
-    // assert(NEQS && "Missing DS node equivalence information.");
-    // auto &Cs = NEQS->getEquivalenceClasses();
-    // auto C = Cs.findValue(representative);
-    // assert(C != Cs.end() && "No equivalence class found.");
-    // assert(Cs.member_begin(C) != Cs.member_end() && "Found empty class.");
-    // if (++(Cs.member_begin(C)) != Cs.member_end()) return false;
-
-    assert(DL && "Missing data layout information.");
-
-    for (auto I = N->type_begin(), E = N->type_end(); I != E; ++I) {
-      if (I->first < offset)
-        continue;
-      if (I->first > offset)
-        break;
-      if (I->second->begin() == I->second->end())
-        break;
-      if ((++(I->second->begin())) != I->second->end())
-        break;
-      Type *T = *I->second->begin();
-      if (!T->isSized())
-        break;
-      if (DL->getTypeAllocSize(T) != length)
-        break;
-      if (!T->isSingleValueType())
-        break;
-      return true;
-    }
-  }
-  return false;
+  return DSA->getNumGlobals(node) == 1 && !node->isArray() &&
```

Comment:
I guess `node` can be nullptr here, and so we should check for that.

Reply:
Actually `node` here can't be nullptr because if it is, `isSingleton` will not be called.